[pipes] PySpark tutorial (#26899)
dagster-io/dagster: docs/content/concepts/dagster-pipes/pyspark.mdx

---
title: "Integrating PySpark with Dagster Pipes | Dagster Docs"
description: "Learn to integrate Dagster Pipes with PySpark to orchestrate PySpark jobs in a Dagster pipeline."
---

# PySpark & Dagster Pipes

This tutorial is focused on using Dagster Pipes to launch & monitor general PySpark jobs. The [Spark integration page](/integrations/spark) provides more information on using Pipes with specific Spark provides (such as AWS EMR or Databricks).

Spark is often used with object stores such as Amazon S3. We are going to simulate this setup locally with [minio](https://min.io/) which has an API compatible with AWS S3. All communication between Dagster and the Spark job (metadata and logs collection) will happen through a Minio bucket.

---

## Prerequisites

> We are going to use `docker` to emulate a typical Spark setup locally. Therefore, only `docker` is required to reproduce this example.

Reviewer: I would format this as just "Docker" and link to the Docker installation page: https://docs.docker.com/get-started/get-docker/

> ## Production Considerations

Reviewer: ```suggestion
## Production considerations
```

> For demonstrational purposes, this tutorial makes a few simplifications that you should consider when deploying in production:

Reviewer: ```suggestion
For demonstration purposes, this tutorial makes a few simplifications that you should consider when deploying in production:
```

> 1. We are going to be running Spark in local mode, sharing the same container with Dagster. In production, consider having two separate environments.

Reviewer: I think this can be formatted as an unordered list, since these aren't steps that need to be taken in sequence.

   - **In the Dagster environment**, you'll need to have the following Python packages:

     - `dagster`
     - `dagster-webserver` --- to run the Dagster UI
     - `dagster-aws` --- when using S3

     And make the orhcestration code available to Dagster (typically via a code location).

   - **In the PySpark environment**, you'll need to install the `dagster-pipes` Python package, and typically the Java AWS S3 SDK packages. For example:

     ```shell
     curl -fSL "https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-aws/3.5.1/hadoop-aws-3.5.1.jar" \
       -o /opt/bitnami/spark/jars/hadoop-aws-3.5.1.jar

     curl -fSL "https://repo1.maven.org/maven2/com/amazonaws/aws-java-sdk-bundle/1.12.780/aws-java-sdk-bundle-1.12.780.jar" \
       -o /opt/bitnami/spark/jars/aws-java-sdk-bundle-1.12.780.jar
     ```

     Make sure `hadoop-aws` JAR and AWS Java SDK bundle versions match your Spark/Hadoop build.

> 2. We are going to upload the PySpark script to the S3 bucket directly inside the Dagster asset. Consider doing it via CI/CD instead.

Reviewer: ```suggestion
2. We are going to upload the PySpark script to the S3 bucket directly inside the Dagster asset. In production, consider doing this via CI/CD instead.
```

---

## Step 1: Writing the Dagster orchestration code

> We will setup a few non-default Pipes componenenbs to streamline the otherwise challenging problem of orchestrating Spark jobs.

Reviewer: ```suggestion
We will set up a few non-default Pipes components to streamline the otherwise challenging problem of orchestrating Spark jobs.
```

> 1. Let's start by creating the asset and opening a Pipes session. We will be using S3 to pass Pipes messages from the Spark job to Dagster. Therefore, we will setup a `PipesS3MessageReader` and `PipesS3ContextInjector` (technically, it's not strictly required to use S3 for passing the Dagster context, but storing it there will decrease the CLI arguments size).

Reviewer: ```suggestion
1. Let's start by creating the asset and opening a Pipes session. We will be using S3 to pass Pipes messages from the Spark job to Dagster, so we will create `PipesS3MessageReader` and `PipesS3ContextInjector` objects. (Technically, it's not strictly required to use S3 for passing the Dagster context, but storing it there will decrease the CLI arguments size.)
```

    s3_client = boto3.client("s3")

    bucket = os.environ["DAGSTER_PIPES_BUCKET"]

    # upload the script to S3
    # ideally, this should be done via CI/CD processes and not in the asset body
    # but for the sake of this example we are doing it here
    s3_script_path = f"{context.dagster_run.run_id}/pyspark_script.py"
    s3_client.upload_file(LOCAL_SCRIPT_PATH, bucket, s3_script_path)

    context_injector = PipesS3ContextInjector(
        client=s3_client,
        bucket=bucket,
    )

    message_reader = PipesS3MessageReader(
        client=s3_client,
        bucket=bucket,
        # the following setting will configure the Spark job to collect logs from the driver
        # and send them to Dagster via Pipes
        include_stdio_in_messages=True,
    )

    with dg.open_pipes_session(
        context=context,
        message_reader=message_reader,
        context_injector=context_injector,
    ) as session:
```

Notice how `PipesS3MessageReader` has `include_stdio_in_messages=True`. This setting will configure the Pipes message writer in the Spark job to collect logs from the Spark driver and send them to Dagster via Pipes messages.

> 2. We will be using CLI arguments to pass the bootstrap information from Dagster to the Spark job. In other Pipes workflows this is typically done via environment variables, but setting them for Spark jobs can be complicated (the exact way of doing this depends on the Spark deployment) or not possible at all. CLI argument is a convenient alternative. We will fetch them from the `session.get_bootstrap_cli_arguments` method. We pass these arguments to `spark-submit` along with a few other settings.

Reviewer: ```suggestion
2. We will be using CLI arguments to pass the bootstrap information from Dagster to the Spark job. In other Pipes workflows this is typically done via environment variables, but setting them for Spark jobs can be complicated (the exact way of doing this depends on the Spark deployment) or not possible at all. CLI arguments are a convenient alternative. We will fetch them from the `session.get_bootstrap_cli_arguments` method. We pass these arguments to `spark-submit` along with a few other settings.
```

---
title: "Integrating PySpark with Dagster Pipes | Dagster Docs"
description: "Learn to integrate Dagster Pipes with PySpark to orchestrate PySpark jobs in a Dagster pipeline."
---

# PySpark & Dagster Pipes

> This tutorial is focused on using Dagster Pipes to launch & monitor general PySpark jobs. The [Spark integration page](/integrations/spark) provides more information on using Pipes with specific Spark provides (such as AWS EMR or Databricks).

Reviewer: ```suggestion
This tutorial is focused on using Dagster Pipes to launch & monitor general PySpark jobs. The [Spark integration page](/integrations/spark) provides more information on using Pipes with specific Spark providers, such as AWS EMR or Databricks.
```

> Spark is often used with object stores such as Amazon S3. We are going to simulate this setup locally with [MinIO](https://min.io/) which has an API compatible with AWS S3. All communication between Dagster and the Spark job (metadata and logs collection) will happen through a MinIO bucket.

Reviewer: ```suggestion
Spark is often used with object stores such as Amazon S3. We are going to simulate this setup locally with [MinIO](https://min.io/), which has an API compatible with AWS S3. All communication between Dagster and the Spark job (metadata and logs collection) will happen through a MinIO bucket.
```

---

## Prerequisites

We are going to use [Docker](https://docs.docker.com/get-started/get-docker/) to emulate a typical Spark setup locally. Therefore, only Docker is required to reproduce this example.

## Production considerations

For demonstration purposes, this tutorial makes a few simplifications that you should consider when deploying in production:

> - We are going to be running Spark in local mode, sharing the same container with Dagster. In production, consider having two separate environments.

Reviewer: ```suggestion
- We are going to be running Spark in local mode, sharing the same container with Dagster. In production, consider having two separate environments:
```

   - **In the Dagster environment**, you'll need to have the following Python packages:

     - `dagster`
     - `dagster-webserver` --- to run the Dagster UI
     - `dagster-aws` --- when using S3

>      And make the orchestration code available to Dagster (typically via a code location).

Reviewer: ```suggestion
     You will also need to make the orchestration code available to Dagster (typically via a code location).
```

Reviewer: I would also recommend making "code location" a link to the code location docs (/concepts/code-locations).

   - **In the PySpark environment**, you'll need to install the `dagster-pipes` Python package, and typically the Java AWS S3 SDK packages. For example:

     ```shell
     curl -fSL "https://repo1.maven.org/maven2/org/apache/hadoop/hadoop-aws/3.5.1/hadoop-aws-3.5.1.jar" \
       -o /opt/bitnami/spark/jars/hadoop-aws-3.5.1.jar

     curl -fSL "https://repo1.maven.org/maven2/com/amazonaws/aws-java-sdk-bundle/1.12.780/aws-java-sdk-bundle-1.12.780.jar" \
       -o /opt/bitnami/spark/jars/aws-java-sdk-bundle-1.12.780.jar
     ```

     Make sure `hadoop-aws` JAR and AWS Java SDK bundle versions match your Spark/Hadoop build.

- We are going to upload the PySpark script to the S3 bucket directly inside the Dagster asset. In production, consider doing this via CI/CD instead.

---

> ## Step 1: Writing the Dagster orchestration code

Reviewer: ```suggestion
## Step 1: Write the Dagster orchestration code
```

> ## Step 1: Writing the Dagster orchestration code

Reviewer: I would call out that you need to create a `dagster_code.py` file to hold the code in this step.

    s3_client = boto3.client("s3")

    bucket = os.environ["DAGSTER_PIPES_BUCKET"]

    # upload the script to S3
    # ideally, this should be done via CI/CD processes and not in the asset body
    # but for the sake of this example we are doing it here
    s3_script_path = f"{context.dagster_run.run_id}/pyspark_script.py"
    s3_client.upload_file(LOCAL_SCRIPT_PATH, bucket, s3_script_path)

    context_injector = PipesS3ContextInjector(
        client=s3_client,
        bucket=bucket,
    )

    message_reader = PipesS3MessageReader(
        client=s3_client,
        bucket=bucket,
        # the following setting will configure the Spark job to collect logs from the driver
        # and send them to Dagster via Pipes
        include_stdio_in_messages=True,
    )

    with dg.open_pipes_session(
        context=context,
        message_reader=message_reader,
        context_injector=context_injector,
    ) as session:
```

Notice how `PipesS3MessageReader` has `include_stdio_in_messages=True`. This setting will configure the Pipes message writer in the Spark job to collect logs from the Spark driver and send them to Dagster via Pipes messages.

> 2. We will be using CLI arguments to pass the bootstrap information from Dagster to the Spark job. In other Pipes workflows this is typically done via environment variables, but setting them for Spark jobs can be complicated (the exact way of doing this depends on the Spark deployment) or not possible at all. CLI arguments are a convenient alternative. We will fetch them from the `session.get_bootstrap_cli_arguments` method. We pass these arguments to `spark-submit` along with a few other settings.

Reviewer: ```suggestion
2. We will be using CLI arguments to pass the bootstrap information from Dagster to the Spark job. In other Pipes workflows, this is typically done via environment variables, but setting them for Spark jobs can be complicated (the exact way of doing this depends on the Spark deployment) or not possible at all. CLI arguments are a convenient alternative. We will fetch them from the `session.get_bootstrap_cli_arguments` method. We pass these arguments to `spark-submit` along with a few other settings.
```

Reviewer: I would reorganize this a bit:

```
We will be using CLI arguments that we fetch from `session.get_bootstrap_cli_arguments` and store in `dagster_pipes_args`. We pass these arguments to `spark-submit`, along with a few other settings, in order to pass the bootstrap information from Dagster to the Spark job.
```

I would move the sentences about environment variables to a note below this step (see my next comment).

# prepare Pipes bootstrap CLI arguments

        dagster_pipes_args = " ".join(
            [
                f"{key} {value}"
                for key, value in session.get_bootstrap_cli_arguments().items()
            ]
        )

        cmd = " ".join(
            [
                "spark-submit",
                # change --master and --deploy-mode according to Spark setup
                "--master",
                "local[*]",
                "--conf",
                "spark.hadoop.fs.s3a.impl=org.apache.hadoop.fs.s3a.S3AFileSystem",
                # custom S3 endpoint for MinIO
                "--conf",
                "spark.hadoop.fs.s3a.endpoint=http://minio:9000",
                "--conf",
                "spark.hadoop.fs.s3a.path.style.access=true",
                f"s3a://{bucket}/{s3_script_path}",
                dagster_pipes_args,
            ]
        )

        subprocess.run(
            # we do not forward stdio on purpose to demonstrate how Pipes collect logs from the driver
            cmd,
            shell=True,
            check=True,  # stdout=subprocess.PIPE, stderr=subprocess.PIPE
        )

        return session.get_results()

> ```

Reviewer: I would add a note here for the environment variables info:

```
<Note>
  In other Pipes workflows, passing this information from Dagster is typically done via environment variables, but setting environment variables for Spark jobs can be complicated (the exact way of doing this depends on the Spark deployment) or not possible at all.
</Note>
```

3. At last, let's add the new asset to `Definitions`:

```python file=/guides/dagster/dagster_pipes/pyspark/dagster_code.py startafter=start_definitions_marker endbefore=end_definitions_marker
defs = dg.Definitions(
    assets=[pipes_spark_asset],
)
```

## Step 2: Use Pipes in the PySpark job

> Call `open_dagster_pipes` in the PySpark job script to create a context that can be used to send messages to Dagster:

Reviewer: I would add a step here to tell the reader to create a `script.py` file to hold the code in this step.

    ) as pipes:
        print("Hello from the Spark driver!")

        pipes.log.info("I am logging a Dagster message from the Spark driver!")

        spark = SparkSession.builder.appName("HelloWorld").getOrCreate()

        df = spark.createDataFrame(
            [(1, "Alice", 34), (2, "Bob", 45), (3, "Charlie", 56)],
            ["id", "name", "age"],
        )

        # calculate a really important statistic
        avg_age = float(df.agg({"age": "avg"}).collect()[0][0])

        # attach it to the asset materialization in Dagster
        pipes.report_asset_materialization(
            metadata={"average_age": {"raw_value": avg_age, "type": "float"}},
            data_version="alpha",
        )

        spark.stop()

if __name__ == "__main__":
    main()
```

Note how `PipesCliArgsParamsLoader` is used to load the CLI arguments passed by Dagster. This information will be used to automatically configure `PipesS3MessageWriter` and `PipesS3ContextLoader`.

Because the Dagster code has `include_stdio_in_messages=True`, the message writer will collect logs from the driver and send them to Dagster via Pipes messages.

## Step 3: Running the pipeline with Docker

> 1. Place the PySpark code to `script.py` and the Dagster orchestration code to `dagster_code.py` in the same directory.

Reviewer: ```suggestion
1. Place the PySpark code for `script.py` and the Dagster orchestration code for `dagster_code.py` in the same directory.
```

# this docker compose file creates a mini Spark cluster with 1 master and 2 workers to simulate a distributed environment

volumes:
  minio-data:
  dagster_home:

networks:
  spark:

services:
  minio:
    image: bitnami/minio
    ports:
      - "9000:9000"
      - "9001:9001"
    environment:
      MINIO_ROOT_USER: minio
      MINIO_ROOT_PASSWORD: minio123
      MINIO_DEFAULT_BUCKETS: "dagster-pipes:public"
    volumes:
      - minio-data:/data
    networks:
      - spark

  dagster-dev:
    develop:
      watch:
        - action: sync
          path: .
          target: /src
    build:
      context: .
      dockerfile: Dockerfile
    command:
      - "dagster"
      - "dev"
      - "-f"
      - "/src/dagster_code.py"
      - "--host"
      - "0.0.0.0"
      - "--port"
      - "3000"
    ports:
      - "3000:3000"
    volumes:
      - spark-logs:/spark/logs
      - spark-data:/spark/data
      - dagster_home:/dagster_home
    environment:
      AWS_ACCESS_KEY_ID: minio
      AWS_SECRET_ACCESS_KEY: minio123
      AWS_ENDPOINT_URL: http://minio:9000
      DAGSTER_PIPES_BUCKET: dagster-pipes

    depends_on:
      - minio

    networks:
      - spark
```

> 4. Start the Dagster dev instance inside docker:

Reviewer: ```suggestion
4. Start the Dagster dev instance inside Docker:
```

```shell
docker compose up --build --watch
```

<Note>
  `--watch` will automatically sync the changes in the local filesystem

>   to the Docker container which is useful for live code updates without volume mounts.

Reviewer: ```suggestion
  to the Docker container, which is useful for live code updates without volume mounts.
```
